### PR TITLE
Fix cloudstack http request client

### PR DESCRIPTION
**Pull request by wassoares**
## Description
This pull request implements the corrections suggested in **issue #451** from the **resource allocation service** project, to avoid throwing a generic exception when receiving an error with the **431** response **status code**.

## Proposed changes
- Fix the **genericRequest** method from the **HttpRequestClient** class, to obtain the error message described in the response headers, refactoring it for a better understanding;
- Make changes to the **callDoGenericRequest** method of the **CloudHttpClient** class, to handle the response headers and obtain a specific description of the error.

## Additional information
- Changed the **package** path of the **HttpRequestClientTest** class, that wasn't in the correct location, causing invisibility of the methods to tests;
- The **CloudStackHttpToFogbowExceptionMapperTest** class, must also have its **package** path corrected;
- No tests were found for the **CloudHttpClient** class;

### CODE REVIEW

File: src/test/java/cloud/fogbow/common/util/connectivity/HttpRequestClientTest.java [hunk 1]
```diff
@@ -0,0 +1,400 @@
...
+
+        PowerMockito.verifyStatic(HttpRequestClient.class, Mockito.times(RUN_ONCE));
+        HttpRequestClient.sendRequestBody(Mockito.eq(connection), Mockito.eq(body));
+
+        PowerMockito.verifyStatic(HttpRequestClient.class, Mockito.times(RUN_ONCE));
+        HttpRequestClient.getHttpResponse(Mockito.eq(connection));
+    }
+
+    // test case: When calling the getHttpResponse method, it must verify that
+    // is call was successful.
```
Comment (jadsonluan) [inline src/test/java/cloud/fogbow/common/util/connectivity/HttpRequestClientTest.java:113]:
```suggestion
    // test case: When calling the getHttpResponse method, it must verify if the call was successful.
```

File: src/test/java/cloud/fogbow/common/util/connectivity/HttpRequestClientTest.java [hunk 2]
```diff
@@ -0,0 +1,400 @@
...
+        PowerMockito.mockStatic(HttpErrorToFogbowExceptionMapper.class);
+        PowerMockito.doReturn(new UnexpectedException())
+                .when(HttpErrorToFogbowExceptionMapper.class, "map", Mockito.anyInt(),
+                Mockito.anyString());
+        try {
+            // exercise
+            HttpRequestClient.getHttpResponse(connection);
+            Assert.fail();
+        } catch (UnexpectedException e) {
+            PowerMockito.verifyStatic(HttpErrorToFogbowExceptionMapper.class, Mockito.times(RUN_ONCE));
```
Comment (jadsonluan) [inline src/test/java/cloud/fogbow/common/util/connectivity/HttpRequestClientTest.java:168]: a `// verify` comment is missing

File: src/test/java/cloud/fogbow/common/util/connectivity/HttpRequestClientTest.java [hunk 3]
```diff
@@ -0,0 +1,400 @@
...
+            HttpErrorToFogbowExceptionMapper.map(Mockito.eq(expectedHttpStatus), Mockito.eq(message));
+        }
+    }
+
+    // test case: When calling the getResponseBody method, it must verify that
+    // is call was successful.
```
Comment (jadsonluan) [inline src/test/java/cloud/fogbow/common/util/connectivity/HttpRequestClientTest.java:174]:
```suggestion
    // test case: When calling the getResponseBody method, it must verify
    // if the call was successful.
```

File: src/test/java/cloud/fogbow/common/util/connectivity/HttpRequestClientTest.java [hunk 4]
```diff
@@ -0,0 +1,400 @@
...
+
+        // verify
+        Mockito.verify(connection, Mockito.times(RUN_ONCE)).getInputStream();
+
+        PowerMockito.verifyStatic(HttpRequestClient.class, Mockito.times(RUN_ONCE));
+        HttpRequestClient.getResponseFrom(Mockito.eq(inputStream));
+    }
+
+    // test case: When calling the getResponseFrom method with a valid input
+    // stream object, it must verify that is call was successful.
```
Comment (jadsonluan) [inline src/test/java/cloud/fogbow/common/util/connectivity/HttpRequestClientTest.java:201]:
```suggestion
    // stream object, it must verify if the call was successful.
```

File: src/test/java/cloud/fogbow/common/util/connectivity/HttpRequestClientTest.java [hunk 5]
```diff
@@ -0,0 +1,400 @@
...
+        PowerMockito.verifyStatic(HttpRequestClient.class, Mockito.times(RUN_ONCE));
+        HttpRequestClient.createBufferedReader(Mockito.eq(inputStreamReader));
+
+        Mockito.verify(bufferedReader, Mockito.times(RUN_ONCE)).readLine();
+        Mockito.verify(bufferedReader, Mockito.times(RUN_ONCE)).close();
+    }
+
+    // test case: When calling the sendRequestBody method with a valid
+    // connection and a body non-empty, it must verify that is call was
+    // successful.
```
Comment (jadsonluan) [inline src/test/java/cloud/fogbow/common/util/connectivity/HttpRequestClientTest.java:234]:
```suggestion
    // connection and a body non-empty, it must verify if the call 
    // was successful.
```

File: src/test/java/cloud/fogbow/common/util/connectivity/HttpRequestClientTest.java [hunk 6]
```diff
@@ -0,0 +1,400 @@
...
+        PowerMockito.verifyStatic(HttpRequestClient.class, Mockito.times(RUN_ONCE));
+        HttpRequestClient.toByteArray(Mockito.eq(body));
+
+        Mockito.verify(outputStream, Mockito.times(RUN_ONCE)).write(Mockito.eq(bytes));
+        Mockito.verify(outputStream, Mockito.times(RUN_ONCE)).flush();
+        Mockito.verify(outputStream, Mockito.times(RUN_ONCE)).close();
+    }
+
+    // test case: When calling the prepareConnection method, it must verify that
+    // is call was successful.
```
Comment (jadsonluan) [inline src/test/java/cloud/fogbow/common/util/connectivity/HttpRequestClientTest.java:266]:
```suggestion
    // test case: When calling the prepareConnection method, it must verify 
    // if the call was successful.
```

File: src/test/java/cloud/fogbow/common/util/connectivity/HttpRequestClientTest.java [hunk 7]
```diff
@@ -0,0 +1,400 @@
...
+        HttpRequestClient.setMethodIntoConnection(Mockito.eq(connection),
+                Mockito.eq(method));
+
+        PowerMockito.verifyStatic(HttpRequestClient.class, Mockito.times(RUN_ONCE));
+        HttpRequestClient.addHeadersIntoConnection(Mockito.eq(connection),
+                Mockito.eq(headers));
+    }
+
+    // test case: When calling the setMethodIntoConnection method, it must
+    // verify that is call was successful.
```
Comment (jadsonluan) [inline src/test/java/cloud/fogbow/common/util/connectivity/HttpRequestClientTest.java:312]:
```suggestion
    // test case: When calling the setMethodIntoConnection method, it must 
    // verify if the call was successful.
```

File: src/test/java/cloud/fogbow/common/util/connectivity/HttpRequestClientTest.java (hunk 2 above)
Comment (wassoares) [reply src/test/java/cloud/fogbow/common/util/connectivity/HttpRequestClientTest.java:169]: Fixed.

File: src/main/java/cloud/fogbow/common/util/connectivity/cloud/CloudHttpClient.java [hunk 8]
```diff
@@ -32,28 +38,55 @@ public String doPostRequest(String url, String bodyContent, T cloudUser)
...
+            }
         }
-
         return response.getContent();
     }
 
+    @VisibleForTesting
+    String getMessageFrom(Map<String, List<String>> responseHeaders) {
+        String message = null;
+        if (responseHeaders.containsKey(CloudStackConstants.X_DESCRIPTION_KEY)) {
```
Comment (fgan1) [inline src/main/java/cloud/fogbow/common/util/connectivity/cloud/CloudHttpClient.java:64]: This "X_Description" key is only used by Cloudstack, isn't it ? And this class is used by Openstack and Cloudstack, rigth? Thus, I think this logic is in the wrong place. Lets discuss it.

File: src/main/java/cloud/fogbow/common/util/connectivity/HttpRequestClient.java [hunk 9]
```diff
@@ -19,55 +25,133 @@
...
         } catch (IOException e) {
             throw HttpErrorToFogbowExceptionMapper.map(responseCode, e.getMessage());
         }
     }
 
-    private static void addHeadersIntoConnection(HttpURLConnection connection, Map<String, String> headers) {
-        for (String key : headers.keySet()) {
-            connection.setRequestProperty(key, headers.get(key));
+    @VisibleForTesting
+    static String getResponseBody(HttpURLConnection connection) {
```
Comment (fgan1) [inline src/main/java/cloud/fogbow/common/util/connectivity/HttpRequestClient.java:56]: Add a @Nulabble annotation.

File: src/main/java/cloud/fogbow/common/util/connectivity/HttpRequestClient.java [hunk 10]
```diff
@@ -19,55 +25,133 @@
...
+        String response = null;
+        try {
+            InputStream inputStream = connection.getInputStream();
+            response = getResponseFrom(inputStream);
+        } catch (IOException e) {
+            LOGGER.error(String.format(Messages.Error.ERROR_MESSAGE_IS_S, e.getMessage()), e);
```
Comment (fgan1) [inline src/main/java/cloud/fogbow/common/util/connectivity/HttpRequestClient.java:62]: You may change the log level to "warn".

File: src/main/java/cloud/fogbow/common/util/connectivity/cloud/CloudHttpClient.java (hunk 8 above)
Comment (wassoares) [reply src/main/java/cloud/fogbow/common/util/connectivity/cloud/CloudHttpClient.java:64]: I believe that we should think about a later refactoring solution for this.

File: src/main/java/cloud/fogbow/common/util/connectivity/HttpRequestClient.java (hunk 9 above)
Comment (wassoares) [reply src/main/java/cloud/fogbow/common/util/connectivity/HttpRequestClient.java:59]: Fixed.

File: src/main/java/cloud/fogbow/common/util/connectivity/HttpRequestClient.java (hunk 10 above)
Comment (wassoares) [reply src/main/java/cloud/fogbow/common/util/connectivity/HttpRequestClient.java]: Fixed.

File: src/test/java/cloud/fogbow/common/util/connectivity/HttpRequestClientTest.java (hunk 1 above)
Comment (wassoares) [reply src/test/java/cloud/fogbow/common/util/connectivity/HttpRequestClientTest.java]: Fixed.

File: src/test/java/cloud/fogbow/common/util/connectivity/HttpRequestClientTest.java (hunk 3 above)
Comment (wassoares) [reply src/test/java/cloud/fogbow/common/util/connectivity/HttpRequestClientTest.java]: Fixed.

File: src/test/java/cloud/fogbow/common/util/connectivity/HttpRequestClientTest.java (hunk 4 above)
Comment (wassoares) [reply src/test/java/cloud/fogbow/common/util/connectivity/HttpRequestClientTest.java]: Fixed.

File: src/test/java/cloud/fogbow/common/util/connectivity/HttpRequestClientTest.java (hunk 5 above)
Comment (wassoares) [reply src/test/java/cloud/fogbow/common/util/connectivity/HttpRequestClientTest.java]: Fixed.

File: src/test/java/cloud/fogbow/common/util/connectivity/HttpRequestClientTest.java (hunk 6 above)
Comment (wassoares) [reply src/test/java/cloud/fogbow/common/util/connectivity/HttpRequestClientTest.java]: Fixed.

File: src/test/java/cloud/fogbow/common/util/connectivity/HttpRequestClientTest.java (hunk 7 above)
Comment (wassoares) [reply src/test/java/cloud/fogbow/common/util/connectivity/HttpRequestClientTest.java]: Fixed.

**Pull request merged**